Add 0.2 & CoffeJam blogpost (#8)
GameDevTecnico/cubos-blog: content/0-2-and-coffeejam.rst

One particular area that is in dire need of development is Audio, which we've been ignoring so far - people expect to hear something when they play a game!
We'll be working on fixing these issues in the next release, and we'll also be profiling the engine to find out where the bottlenecks are.

New 0.2 Tools Features
======================

Transform Gizmo Upgrades :dim:`(@DiogoMendonc-a)`
-------------------------------------------------

The biggest change was the addition of a rotation gizmo, so now you can rotate entities using your mouse!

We also added a toggle that allows changing between using global or local space with the Transform Gizmo,
and a new type of gizmo, a ``Rotated Box``, which, unlike the old boz gizmos, does not need to be axis-aligned.

Finally, the transform gizmo is now always rendered at the same size, regardless of the selected entity's distance to the camera.
This should hopefully make it easier to use the tool when moving entities either very far away, or very close.

.. image:: images/transform_gizmo.gif

World Inspector Overhaul :dim:`(@diogomsmiranda)`
-------------------------------------------------

> The World inspector as been on our radar for quite some time now, as a tool that could be improved.

Reviewer: ```suggestion
The World inspector has been on our radar for quite some time now as a tool that could be improved.
```

This release we particularly aimed on making it better by focusing on:

> * Making it easier to find the entities you are looking for.

Reviewer: ```suggestion
* Making it easier to find the entities you are looking for;
```

> * Getting more information about the entities in the scene.

Reviewer: ```suggestion
* Displaying more information about the entities in the scene.
```

> To tackle the first point I've added a search bar that allows you to filter the entities in the scene, either by their name, or by their components. i.e:

Reviewer: ```suggestion
To tackle the first point we've added a search bar that allows you to filter the entities in the scene, either by their name, or by their components. i.e:
```
Should we always use the plural, for consistency?

> To tackle the first point I've added a search bar that allows you to filter the entities in the scene, either by their name, or by their components. i.e:

Reviewer: ```suggestion
To tackle the first point I've added a search bar that allows you to filter the entities in the scene, either by their name or by their components. e.g.:
```

* Searching for "player" will show all entities that have "player" in their name.
* Searching for "Transform" will show all entities that have a ``Transform`` component.
* Searching for "player, Transform" will show all entities that have a ``Transform`` component and contain "player" in their name.

To tackle the second point on our list we added an hierarchy view that shows the entities in the scene in a tree-like structure representing the ``ChildOf`` relation between entities.

Sub-scene importing in scene editor :dim:`(@teres4)`
-----------------------------------------------------

Previously our scene editor was lacking support for importing sub-scenes.
This meant that developers had to manually open the scene file and paste the sub-scene UUID in its import section.
To save time, I've added a popup that shows all the importable sub-scenes, and allows the developer to select the one they want to import.

New 0.2 Engine Features
=======================

> Our focus on this release was to continue working on improving and implementing base functionality, such as rendering and physics, and also to improve on the tools available to the developer.

Reviewer: Shouldn't this be at the top?

We've compiled here some of the most important changes (if you're curious, you can check out the full changelog in our `repository <https://github.com/GameDevTecnico/cubos/blob/main/CHANGELOG.md>`_).

Physics Improvements :dim:`(@fallenatlas)`
------------------------------------------

> This update we introduced the ``Solver``, which works to satisfy the physics constraints applied to entities (eg. Spring, Fixed distance between entities, etc.).

Reviewer: ```suggestion
This update we introduced the ``Solver``, which works to satisfy the physics constraints applied to entities (e.g. Spring, Fixed distance between entities, etc.).
```

To allow repeating groups such as these, I added a new feature to the ECS tags: the repeat while modifier. This makes the systems tagged by it repeat until a certain condition becomes false.

.. code-block:: cpp

    cubos.tag(groupTag).repeatWhile(myCondition);
    cubos.system("integrate physics").tagged(groupTag)./* ... */;
    cubos.system("update velocity").tagged(groupTag)./* ... */;
    cubos.system("clear forces").tagged(groupTag)./* ... */;

Its also possible to make repeating subgroups by tagging the subgroup with the parent tag.

.. code-block:: cpp

    cubos.tag(subgroupTag).tagged(groupTag).repeatWhile(myOtherCondition);

This way the subgroup's systems will repeat a total of n x m times, where n is the number of times the parent tag repeats and m is the number of times the subgroup tag repeats.

With this implemented, it was easy to implement the fixed-step plugin on the engine side. It simply adds a tag that makes systems repeat as needed according to
the ``DeltaTime`` passed, avoiding variance due to different frame rates and more/less powerful PCs.

Moving Away from the Old Serialization :dim:`(@Dacops)`
-------------------------------------------------------

> `Serialization <https://gamedevtecnico.github.io/cubos/docs/group__core-data-ser.html>`_ is a crucial part of the game engine that allows for the saving of any CUBOS. game components as well as then loading them in.

Reviewer: ```suggestion
`Serialization <https://gamedevtecnico.github.io/cubos/docs/group__core-data-ser.html>`_ is a crucial part of the game engine that allows for the saving of any CUBOS game components as well as then loading them in.
```

> Previously, serialization was done through a series of functions, both for reading and writing that could be overloaded for each type.

Reviewer: ```suggestion
Previously, serialization was done through a series of functions, both for reading and writing, that could be overloaded for each type.
```
